Fix bug with Trilinos and multiple resistance BCs

aabrown100-git opened the pull request:
<!-- Give your pull request (PR) a descriptive name -->

## Current situation
Resolves https://github.com/SimVascular/svMultiPhysics/issues/259


## Release Notes 
- Fix error how Trilinos computes the resistance BC contributions to the tangent matrix when multiple resistance BCs are applied (see [here](https://github.com/SimVascular/svMultiPhysics/issues/259#issuecomment-2837061558) for detailed explanation).
- Added iliac artery with Trilinos linear solver test case 

## Testing
Added test case using Trilinos linear solver to solve existing iliac artery test case. This case has two resistance BCs and so is appropriate to test this bug fix. Before bug fix, this case show very poor nonlinear convergence. After bug fix, nonlinear convergence is practically identical to the iliac artery case using the FSILS GMRES linear solver.


## Code of Conduct & Contributing Guidelines 
- [ ] I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

ktbolt commented on Code/Source/solver/trilinos_impl.cpp:
```diff
@@ -100,32 +100,41 @@ bool coupledBC;
-    //now need to add on bdry term v*(v'*x)
-    double *dot = new double[1]; //only 1 multivector to store result in it
+    // Declare dot product v_i'*x
+    double *dot = new double[1];
```
Why not declare `dot`  `double dot[1]` ?

aabrown100-git replied:
Changed

ktbolt commented on Code/Source/solver/trilinos_impl.cpp:
```diff
@@ -943,14 +965,23 @@ void trilinos_bc_create_(const double *v, bool &isCoupledBC)
+                &v[i*dof]); //values of size dof
+        if (error != 0)
+        {
+          std::cout << "ERROR: Setting boundary vector values!" << std::endl;
```
Throw an exception here.

aabrown100-git replied:
Will do

ktbolt commented on Code/Source/solver/trilinos_impl.cpp:
```diff
@@ -1208,25 +1241,34 @@ void TrilinosLinearAlgebra::TrilinosImpl::init_dir_and_coup_neu(ComMod& com_mod,
     }
+
+    // Add the coupled Neumann BC vector to the list
+    v_list.push_back(v);
```
This will perform a copy of `v`.

aabrown100-git replied:
Thanks for pointing out. I changed it so that no copies are made

ktbolt commented on Code/Source/solver/trilinos_impl.h:
```diff
@@ -189,7 +189,7 @@ class TrilinosMatVec: public virtual Epetra_Operator
           int& numGhostAndLocalNodes, int& nnz, const int *ltgSorted,
           const int *ltgUnsorted, const int *rowPtr, const int *colInd,
-          int &dof, int& cpp_index, int& proc_id);
+          int &dof, int& cpp_index, int& proc_id, int& numCoupledNeumannBC);
```
Why are all of these int parameters being declared as references, are they modified in this function ?

aabrown100-git replied:
Not sure why they are declared as references. Most if not all do not appear to be modified by this function

ktbolt replied:
@aabrown100-git This a left-over from svFSI where these C functions were called from Fortran.

aabrown100-git replied:
Thanks @ktbolt, would you like us to change it now?

ktbolt replied:
@aabrown100-git It is good to clean up things like this when we encounter them, reduces future confusion. 

Please change all of these `int &` to `const int`. 

And if you are a righteous coder

- The `int *` parameters are raw pointers from `Vector<int>` objects so these guys could also be changed to `Vector<int>&` but you will have the also modify the `trilinos_lhs_create_()` code in a couple of places. 

- Remove the `_` from `trilinos_lhs_create_()`.

aabrown100-git replied:
Thanks for the suggestions, just made all of these changes.

aabrown100-git commented on Code/Source/solver/trilinos_impl.cpp:
```diff
@@ -943,14 +965,23 @@ void trilinos_bc_create_(const double *v, bool &isCoupledBC)
+        auto v = v_list[k].data();
+        auto bdryVec = Trilinos::bdryVec_list[k];
+        //set the coupled boundary vector
+        error = bdryVec->ReplaceGlobalValues(
```
I think this comment is incorrect

aabrown100-git commented on Code/Source/solver/trilinos_impl.cpp:
```diff
@@ -943,14 +965,23 @@ void trilinos_bc_create_(const double *v, bool &isCoupledBC)
+        auto v = v_list[k].data();
+        auto bdryVec = Trilinos::bdryVec_list[k];
+        //set the coupled boundary vector
+        error = bdryVec->ReplaceGlobalValues(
```
Wrong, error is checked

aabrown100-git commented on Code/Source/solver/trilinos_impl.cpp:
```diff
@@ -100,32 +100,41 @@ bool coupledBC;
-    Trilinos::bdryVec->Dot(x,dot);
+    // Loop over all coupled Neumann boundary vectors
+    for (auto bdryVec : Trilinos::bdryVec_list)
+    {
+      // Compute dot product dot = v_i'*x
+      bdryVec->Dot(x, dot);
 
-    //Y = 1*Y + dot*v daxpy operation
-    y.Update(*dot, //scalar for v
-             *Trilinos::bdryVec, //FE_Vector
-             1.0); ///scalar for Y
+      // y = 1*y + dot*v_i
+      y.Update(*dot,
+              *bdryVec,
+              1.0);
+    }
 
     delete[] dot;
```
Changed

ktbolt reviewed (approved):
Clean and clear, very nice!

aabrown100-git commented:
@divyaadil23 and I added a new fluid Trilinos test case, but we forgot to put it in test_fluid.py so it wasn't actually being tested! Fixed this in the most recent commit.
